### PR TITLE
Add Spill To Disk For ORDER BY

### PR DESCRIPTION
**Pull request by atris**
ORDER BY currently will error out if the data being processed exceeds
query memory limit. The PR introduces paging from disk and ensures
that ORDER BY is limited only by the amount of disk present

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 1]
```diff
@@ -62,6 +74,13 @@ public OrderByOperatorFactory(
             this.sortOrder = ImmutableList.copyOf(requireNonNull(sortOrder, "sortOrder is null"));
 
             this.pagesIndexFactory = requireNonNull(pagesIndexFactory, "pagesIndexFactory is null");
+            this.spillEnabled = spillEnabled;
+
+            if (spillEnabled) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: Change to:
```
checkState(!spillEnabled || spillerFactory.isPresent(), "...
```

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 2]
```diff
@@ -104,12 +125,19 @@ public OperatorFactory duplicate()
...
     private final List<SortOrder> sortOrder;
     private final int[] outputChannels;
+    private final LocalMemoryContext localRevocableMemoryContext;
     private final LocalMemoryContext localUserMemoryContext;
 
     private final PagesIndex pageIndex;
 
-    private final PageBuilder pageBuilder;
-    private int currentPosition;
+    private List<Type> sourceTypes;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: source types can be final

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 3]
```diff
@@ -104,12 +125,19 @@ public OperatorFactory duplicate()
...
+
+    private final boolean spillEnabled;
+    private final Optional<SpillerFactory> spillerFactory;
+    private Optional<Spiller> spiller = Optional.empty();
+    private ListenableFuture<?> spillInProgress = immediateFuture(null);
+
+    private WorkProcessor<Page> mergedPages;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: you could just keep `Iterator<Page>` here create via: `WorkProcessor#iterator`

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: Rename to `sortedPages`

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 4]
```diff
@@ -147,8 +180,7 @@ public void finish()
         if (state == State.NEEDS_INPUT) {
             state = State.HAS_OUTPUT;
 
-            // sort the index
-            pageIndex.sort(sortChannels, sortOrder);
+            mergedPages = unspill();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: this suggest that spill happened. Maybe call `unspill()`:
```
getSortedPages()
```
?

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 5]
```diff
@@ -218,4 +295,60 @@ public void close()
...
+                requireNonNull(pageWithPositionComparator, "comparator is null"),
+                sourceTypes,
+                operatorContext.aggregateUserMemoryContext(),
+                operatorContext.getDriverContext().getYieldSignal());
+    }
+
+    private void updateMemoryUsage(boolean useRevocableMemory)
+    {
+        if (useRevocableMemory) {
+            if (!spillEnabled) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: `checkState(spillEnabled, "Revocable memory requested when spill to disk is not enabled"); `

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 6]
```diff
@@ -185,23 +213,72 @@ public Page getOutput()
             return null;
         }
 
-        if (currentPosition >= pageIndex.getPositionCount()) {
+        if (!mergedPages.process() || mergedPages.isFinished()) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: `!mergedPages.process()` doesn't mean that process finished.

Convert processor to iterator and it should be easier to use in this context.

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 7]
```diff
@@ -120,19 +148,24 @@ public OrderByOperator(
...
+            Optional<SpillerFactory> spillerFactory)
     {
         requireNonNull(pagesIndexFactory, "pagesIndexFactory is null");
 
         this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
         this.outputChannels = Ints.toArray(requireNonNull(outputChannels, "outputChannels is null"));
         this.sortChannels = ImmutableList.copyOf(requireNonNull(sortChannels, "sortChannels is null"));
         this.sortOrder = ImmutableList.copyOf(requireNonNull(sortOrder, "sortOrder is null"));
+        this.sourceTypes = sourceTypes;
         this.localUserMemoryContext = operatorContext.localUserMemoryContext();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: rename those to:
```
pagesIndexUserMemoryContext
pagesIndexRevocableMemoryContext
```

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 8]
```diff
@@ -185,23 +213,72 @@ public Page getOutput()
...
-        currentPosition = pageIndex.buildPage(currentPosition, outputChannels, pageBuilder);
+        Page resultPage = mergedPages.getResult();
+        Block[] blocksArray = new Block[outputChannels.length];
 
-        // output the page if we have any data
-        if (pageBuilder.isEmpty()) {
-            state = State.FINISHED;
-            return null;
+        // Do not use revocable memory during unspill
+        updateMemoryUsage(false);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: that shouldn't be needed

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 9]
```diff
@@ -185,23 +213,72 @@ public Page getOutput()
...
-        if (currentPosition >= pageIndex.getPositionCount()) {
+        if (!mergedPages.process() || mergedPages.isFinished()) {
             state = State.FINISHED;
             return null;
         }
 
-        // iterate through the positions sequentially until we have one full page
-        pageBuilder.reset();
-        currentPosition = pageIndex.buildPage(currentPosition, outputChannels, pageBuilder);
+        Page resultPage = mergedPages.getResult();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: why not return page directly?

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 10]
```diff
@@ -185,23 +213,72 @@ public Page getOutput()
...
+            return spillInProgress;
+        }
+
+        pageIndex.sort(sortChannels, sortOrder);
+        spillInProgress = spiller.get().spill(pageIndex.getSortedPages());
+
+        return spillInProgress;
+    }
+
+    private boolean hasPreviousSpillCompletedSuccessfully()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: nit: I'm not convinced we still need those, but we could keep them

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 11]
```diff
@@ -185,23 +213,72 @@ public Page getOutput()
...
+    public void finishMemoryRevoke()
+    {
+        if (pageIndex.getPositionCount() == 0) {
+            return;
         }
 
-        Page page = pageBuilder.build();
-        return page;
+        pageIndex.clear();
+        localRevocableMemoryContext.setBytes(pageIndex.getEstimatedSize().toBytes());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: `pageIndex` memory overhead is non-revokable, see: https://github.com/prestodb/presto/pull/11364/files#diff-f0a46746520bef39a6a15ddb4e65aa56R607

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 12]
```diff
@@ -218,4 +295,60 @@ public void close()
...
+        }
+
+        return spiller.get().getSpills().stream()
+                .map(WorkProcessor::fromIterator)
+                .collect(Collectors.toList());
+    }
+
+    public WorkProcessor<Page> mergeSpilledPagesAndCurrentPagesIndex(PagesIndex currentPagesIndex, List<WorkProcessor<Page>> spilledPages)
+    {
+        PageWithPositionComparator pageWithPositionComparator = new SimplePageWithPositionComparator(sourceTypes, sortChannels, sortOrder);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: this should be compiled as for spilling window operator

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 13]
```diff
@@ -218,4 +295,60 @@ public void close()
...
+        WorkProcessor<Page> pageIndexPages = WorkProcessor.fromIterator(currentPagesIndex.getSortedPages());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: Please shortcircut the case when spill did not happen. In such case we don't need to merge pages index rows, but can use `PagesIndex#getSortedPages` directly. That should be faster

File: presto-main/src/test/java/com/facebook/presto/operator/DummySpillerFactory.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+
+public class DummySpillerFactory
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/operator/DummySpillerFactory.java]: you could make it a separate commit

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java (hunk 11 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: Good catch, thanks

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java (hunk 13 above)
Comment (atris) [reply presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: That is what happens today in ``unspill``. We will return an ``iterator`` of the pages in the pagesIndex, by calling ``getSortedPages`` directly

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 15]
```diff
@@ -185,23 +216,69 @@ public Page getOutput()
...
-        pageBuilder.reset();
-        currentPosition = pageIndex.buildPage(currentPosition, outputChannels, pageBuilder);
+        Page resultPage = sortedPages.next();
+        Block[] blocksArray = new Block[outputChannels.length];
 
-        // output the page if we have any data
-        if (pageBuilder.isEmpty()) {
-            state = State.FINISHED;
-            return null;
+        for (int i = 0; i < outputChannels.length; i++) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: you can return `resultPage` directly

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 16]
```diff
@@ -185,23 +216,69 @@ public Page getOutput()
...
+            blocksArray[i] = resultPage.getBlock(outputChannels[i]);
+        }
+
+        return new Page(resultPage.getPositionCount(), blocksArray);
+    }
+
+    @Override
+    public ListenableFuture<?> startMemoryRevoke()
+    {
+        return spillToDisk();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: inline `spillToDisk` method

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 17]
```diff
@@ -185,23 +216,69 @@ public Page getOutput()
...
+        pageIndex.sort(sortChannels, sortOrder);
+        spillInProgress = spiller.get().spill(pageIndex.getSortedPages());
+
+        return spillInProgress;
+    }
+
+    private boolean hasPreviousSpillCompletedSuccessfully()
+    {
+        if (spillInProgress.isDone()) {
+            // check for exception from previous spill for early failure
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: I don't get it. `getFutureValue(spillInProgress);` might still might not be called if there is no next spill

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 18]
```diff
@@ -147,8 +180,10 @@ public void finish()
         if (state == State.NEEDS_INPUT) {
             state = State.HAS_OUTPUT;
 
-            // sort the index
-            pageIndex.sort(sortChannels, sortOrder);
+            WorkProcessor<Page> resultPages = getSortedPages(
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: Here we should also switch accounting from revokable to non-revokable as we don't support spilling when producing output (yet), but such allocation might fail then. We probably should do similar as in: https://github.com/prestodb/presto/pull/11654/files#r226286007

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: @findepi I think `checkSpillCompletedSuccessfully` should actually be part of `finishMemoryRevoke`. The reason is that we might voluntarily start spilling (as in https://github.com/prestodb/presto/pull/11654/files#r226286007), but then if `startMemoryRevoke` gets called we might be in a middle of such voluntarily spilling

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java [hunk 19]
```diff
@@ -210,12 +287,61 @@ public void close()
         pageIndex.clear();
     }
 
-    private static List<Type> toTypes(List<? extends Type> sourceTypes, List<Integer> outputChannels)
+    public WorkProcessor<Page> getSortedPages(PageWithPositionComparator pageWithPositionComparator)
+    {
+        List<WorkProcessor<Page>> spilledPages = getSpilledPages();
+
+        if (spilledPages.isEmpty()) {
+            pageIndex.sort(sortChannels, sortOrder);
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: This is also an issue for window spill. We construct pages index there which might be over `quer-max-memory-per-node`. In such case we should go through spill path (for now at least, we could discard processed partitions as an improvement later on)

File: presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java]: > I don't get it. `getFutureValue(spillInProgress);` might still not be called if there is no next spill, so previous spill failure might never get checked.

Good point. We should also add such a check later, eg. in getinput or getoutput

> @findepi I think checkSpillCompletedSuccessfully should actually be part of finishMemoryRevoke

Should be as well, yes.
However, i would like to maintain the convention where you check `spillInProgress`'s result whenever you overwrite this with a new future.

**Comment by findepi** [comment]
Superseded by https://github.com/prestodb/presto/pull/11744
